virtual_rainforest/core/config.py

```diff
@@ -236,6 +233,80 @@ def load_in_config_files(files: list[Path]) -> dict:
     return config_dict
 
 
+def extend_with_default(
+    validator_class: type[Draft202012Validator],
+) -> type[Draft202012Validator]:
+    """Extend validator so that it can populate default values from the schema.
+
+    Args:
+        validator_class: Validator to be extended
+    """
+
+    validate_properties = validator_class.VALIDATORS["properties"]
+
+    def set_defaults(
+        validator: type[Draft202012Validator],
+        properties: dict,
+        instance: dict,
+        schema: dict,
+    ) -> Iterator:
+        """Generate an iterator to populate defaults."""
+        for property, subschema in properties.items():
+            if "default" in subschema:
+                instance.setdefault(property, subschema["default"])
+
+        for error in validate_properties(
+            validator,
+            properties,
+            instance,
+            schema,
+        ):
+            yield error
+
+    return validators.extend(
+        validator_class,
+        {"properties": set_defaults},
+    )
+
+
+def add_core_defaults(config_dict: dict) -> dict:
```

Comment:
I have a feeling that you could restrict your `dict` types somewhat more by specifying the key type: `dict[str, any]`

There are some other places where I think you could do this too

Reply:
Ahh right yes I wasn't sure how to type nested dictonaries, but `dict[str, any]` seems sensible